man/vznnc.8.in
.TH vznnc 8 "29 May 2014" "OpenVZ" "Containers"
.SH NAME
vznnc \- run a program connected to a socket
.SH SYNOPSIS
.SY vznnc
{
.B -l
|
.B -c
}
.B -p
.I port
.OP -f fd
.OP --
.I program
[\fIarg\fR \fI...\fR]
.SH DESCRIPTION
This "nano-netcat" utility can be used to either listen on or connect to
a TCP port at localhost, and run a specified \fIprogram\fR with its stdin
and stdout (or a specified file descriptor) connected to the socket.

.SH OPTIONS
.TP
.B -l
Listen on a specified TCP port at localhost.
.TP
.B -c
Connect to a specified TCP port at localhost.
.TP
\fB-p\fR \fIport\fR
Port number.
.TP
\fB-f\fR \fIfd\fR
File descriptor ID. If this option is not set, stdin and stdout are closed
and are connected to the socket, otherwise they are left intact, and the
specified \fIfd\fR is used.
.TP
.B --
This is a separator between
.B vznnc
own arguments and
.I program
arguments, so that latter won't be processed by
.BR vznnc .
It is required in case there are any arguments to
.I program
that start with the dash
.RB ( - )
character, and is optional otherwise.
.TP
\fIprogram\fR [ \fIarg\fR \fI...\fR ]
Program to run, with optional arguments.
.SH EXIT STATUS
Returns \fIprogram\fR exit status upon success, or one of the following
codes in case of an error:
.IP 1
Invalid usage
.IP 127
Error executing \fIprogram\fR.
.IP 220
Network-related error.
.SH EXAMPLES
To run receiving side of
.B ploop copy
command on a remote server, using openssh port forwarding:

.EX
 PORT=2345
 ssh -L localhost:$PORT:localhost:$PORT $REMOTE_SERVER \\
	vznnc -l -p $PORT -- ploop copy -d $FILE -i0 -o1
.EE

To do the same, but with stdin and stdout intact, using file descriptor 5
for communication:

.EX
 ssh -L localhost:$PORT:localhost:$PORT $REMOTE_SERVER \\
	vznnc -l -p $PORT -f 5 -- ploop copy -d $FILE -i5 -o5
.SH SEE ALSO
.BR nc (3),
.BR netcat (3),
.BR socat (2).
.SH LICENSE
Copyright (C) 2014, Parallels, Inc. Licensed under GNU GPL v2.


ChangeLog
Since after 3.0.12 release, this file is obsoleted. Complete change log
can be obtained from OpenVZ git repository.

	http://git.openvz.org/?p=vzctl;a=shortlog
	http://git.openvz.org/?p=vzctl;a=log

Below is a change log for old vzctl releases.


   3.0.12: Thu Oct 5 2006 <igor@openvz.org>
	* vzctl start: added --wait cmdline option
	* added vznetcfg script for veth setup
	* fixed error message in case kernel compiled without USER_RESOURCES
	* added '--applyconfig_map name' vzctl set option
	* changed friendly name data structure
	* fixed vzmigrate to support VE name feature (bug #239)
	* fixed annoying arp warning (bug #189)
	* added bash completion script (bug #212)
	* added check to vzmigrate if OpenVZ is run on another side (bug #163)
	* removed debian directory from the tarball (bugs #197, #194)
	* removed LICVIEW from /etc/init.d/vz (bug #237)
	* fixed misprints and /dev/nul(l) usage in dist scripts (bug #228)
	* fixed vzmigrate to work with different awk flavors (bug #231)
	* years in copyright fixed (bug #232)
	* increased some UBCs in ve-light.conf-sample (bug #234)
	* fixed some minor memory leaks while doing realloc()
	* vzctl destroy: remove VE names on destroy (bug #236)
	* added --name option description to vzctl(8) man page (bug #232)
	* merged libvzctl-simfs and libvzctl into a single libvzctl
	* vzlist: replaced sprintf() with snprintf() to avoid buffer overflows
	* vzctl enter: forward error messages from child process
	* vzctl enter: fix off-by-one error which resulted in
	  uninitialized PATH
	* arpsend: fixed segfault when parsing long options
	* vzlist: fixed cast from pointer to integer of different size warnings
	* vzctl enter: exit with correct error code on enter fail (bug #245)
	* added udev rules for /dev/vzctl (bug #204)
	* added syscall numbers for powerpc
	* /etc/init.d/vz: check for /dev/vzctl only after modprobe (bug #204)
	* fixed ppc64 compilation
	* removed distro-specific stuff from 'make install' (bug #254)
	* added 'make install-gentoo' and Gentoo init.d/vz script (bug #262)
	* fixed arpsend options parsing for some compilers (bug #261)
	* man pages cleanup (VPS->VE, etc.)
	* changed all utilities and man pages to refer to VE not VPS
	* vzcpucheck: give error if VE0 CPU units not set (bug #249)

   3.0.11: Mon Aug  7 2006 <igor@openvz.org>
	* fixed using VE_ROUTE_SRC_DEV variable
	* added VPS name support
	* added change number of CPUs available in the VPS functionality.
	* fixes in CPT error processing
	* added upport of xt_mac iptables module
	* fixed awk expression to be compatible with mawk

   3.0.10: Wed Jun  7 2006 <igor@openvz.org>
	* added virtual ethernet device support (--veth_add, --veth_del)
	* added  /proc/meminfo virtualization support (--meminfo mode:value)
	* added VE0 UB setup functionality (--reset-ub)

   3.0.9: Wed May 24 2006 <igor@openvz.org>
	vzctl:
	* switched to new versioning scheme.
	* added -z option to /sbin/init
		(http://bugzilla.openvz.org/show_bug.cgi?id=164)
	* reset standart file desc [0-2] to /dev/null in case they are closed.
	* added possibility to set cpulimit in %
	* added suffixes G,M,K support for --diskspace option
	* set SIGPIPE handler to SIG_IGN before fist write to stdout/err
	* added Debian packaging support
	* changed directory structure.
		/etc/sysconfig/vz-scripts -> /etc/vz/conf
		/etc/sysconfig/vz-scripts/dists -> /etc/vz/dists
		/etc/sysconfig/vz -> /etc/vz/vz.conf
	* better distribution detection in vz initscript.
	* added Ubuntu distribution configuration file as copy of debian.
		(http://bugzilla.openvz.org/show_bug.cgi?id=119)
	* Increased kmemsize, privmpages, numfiles for x86_64
		ve-vps.basic.conf-sample.
		(http://bugzilla.openvz.org/show_bug.cgi?id=69)
	* changed -p -> --preserve=mode,ownership for 'cp' workaround in
		case /proc is not mounted.
	* internal modifications
		changed dynamic UB structure to static one.
		moved VPS configure functionality to vps_configure.c
		moved destroy functionaliny to destroy.c
	vzmigrate:
	* move config as veid.migrated to allow backward migration
		(http://bugzilla.openvz.org/show_bug.cgi?id=152)
	* added coping of action scripts.
		(http://bugzilla.openvz.org/show_bug.cgi?id=157)
	* fixed --keep-dst option processing. Option behavior was inverted.

   3.0.0-8: Mon Apr 24 2006 <igor@openvz.org>
	* UBC parameters > LONG_MAX in VPS conf truncate to LONG_MAX
	* added installation of vzmigrate.8 man page
	* use 'uname -m' for arch selection

   3.0.0-6: Tue Apr 18 2006 <kir@openvz.org>
	* man page fixes
   3.0.0-5: Fri Apr  7 2006 <igor@sw.ru>
	* added checkpointing support
	* added kernel headers for standalone compilation
	* chaged license from QPL to GPL
	* set personality PER_LINUX32 on x86_64 arch for i386 VPSes
	* fixed --iptables store (do not store dependent modules)
	* Do skip unknown iptables modules in config
	* set arg[0] = "-bash" on enter

   3.0.0-4: Thu Mar 16 2006 <kir@openvz.org>
	* cron scripts do nothing if VZ is not loaded (bug #107)
	* cosmetic vzctl.8 man page fixes
	* reverted bug #110 changes, see http://bugzilla.openvz.org/111
	* added ipt_REDIRECT support, thanks to jstubbs-at-gentoo.org

   3.0.0-3: Thu Mar  2 2006 <igor@sw.ru>
	* put_param3() can incorrectly set $name=( "" ) (bug #111).
		Affects gentoo VPSs. Thanks to jstubbs-at-gentoo.org
	* fixed compilation against old kernels (<= 022stab070)
	* description of set parameter suffixes (K, M, G, P) added
		to vzctl.8 man page (bug #105)
	* fixed gentoo-add_ip.sh script (bug #110)

   3.0.0-2: Thu Feb 23 2006 <igor@sw.ru>
	* fixed VPS start on old kernel (witthout feature_mask support)

   3.0.0-1: Wed Feb 22 2006 <igor@sw.ru>
	* enable sysfs for SuSE like distributions
	* added possibility to setup UBC parameters in the form
		xxx[GMKPB]:yyy[GMKPB]

   2.7.0-27: Tue Feb  7 2006 <igor@sw.ru>
	* fixed VE_CONFFILE variable passed to vps.mount script
	* fixed venet0 initialization in vz initscript (bug #99)

   2.7.0-26: Thu Feb  2 2006 <igor@sw.ru>

src/vznnc.c
/*
 *  Copyright (C) 2014, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* vzncc: VZ Nano NetCat, a trivial tool to either listen or connect to
 * a TCP port at localhost, and run a program with its input and output
 * redirected to the connected socket
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <sys/poll.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#define FAIL 220
#define SELF "vznnc"

void usage(void) {
	fprintf(stderr,
		"Usage: " SELF " {-l|-c} -p PORT [-f FD] CMD [arg ...]\n");

	exit(1);
}

/* Wait on socket before accept(), while checking that stdin/stdout
 * are not gone. This helps to solve a problem of stale vznnc
 * that was run by ssh if ssh is terminated.
 *
 * Return:
 *   0 - time to do accept()
 *   1 - stdin/stdout are gone, time to exit
 *  -1 - some unexpected error
 */
int pollwait(int fd) {
	struct pollfd pfd[] = {
		{.fd = fd, .events = POLLIN},
		{.fd = 0},
		{.fd = 1},
	};
	const int ev_err = POLLERR|POLLHUP;
	int ret;

	do {
		ret = poll(pfd, sizeof(pfd) / sizeof(pfd[0]), 10 * 1000);
		if (ret < 0) {
			perror(SELF ": poll()");
			return -1;
		}
	} while (ret == 0);

	if (pfd[0].revents & POLLIN)
		/* ready for accept() */
		return 0;

	if (pfd[1].revents & ev_err || pfd[2].revents & ev_err) {
		/* stdin or stdout closed */
		/* most probably stderr is gone too, but let's try */
		fprintf(stderr, SELF ": broken pipe!");
		return 1;
	}

	/* some other unknown event, treat as error */
	return -1;
}

int main(int argc, char *argv[])
{
	int sockfd, connfd, port = -1;
	struct sockaddr_in srv_addr = {};
	int lis = 0, con = 0, fd = -1;
	const int yes = 1;

	while (1) {
		int c;

		c = getopt(argc, argv, "lcp:f:");
		if (c == -1)
			break;
		switch (c) {
			case 'l':
				lis++;
				break;
			case 'c':
				con++;
				break;
			case 'p':
				port = atoi(optarg);
				break;
			case 'f':
				fd = atoi(optarg);
				break;
			default:
				usage();
		}
	}

	if (lis && con) {
		fprintf(stderr, SELF ": options -l and -c are exclusive\n");
		usage();
	}

	if (!lis && !con) {
		fprintf(stderr, SELF ": either -l or -c is required\n");
		usage();
	}

	if (port < 0) {
		fprintf(stderr, SELF ": option -p is required\n");
		usage();
	}

	if (argc - optind < 1)
		usage();

	sockfd = socket(AF_INET, SOCK_STREAM, 0);
	if (sockfd < 0) {
		perror(SELF ": socket()");
		return FAIL;
	}

	if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int))) {
		perror(SELF ": setsockopt()");
		return FAIL;
	}

	srv_addr.sin_family = AF_INET;
	srv_addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
	srv_addr.sin_port = htons(port);

	if (lis) { /* listen */
		struct sockaddr_in cl_addr = {};
		socklen_t cl_len = sizeof(cl_addr);

		if (bind(sockfd, (struct sockaddr *) &srv_addr,
					sizeof(srv_addr)) < 0) {
			perror(SELF ": bind()");
			return FAIL;
		}

		if (listen(sockfd, 0) < 0) {
			perror(SELF ": listen()");
			return FAIL;
		}

		if (pollwait(sockfd))
			return FAIL;

		connfd = accept(sockfd, (struct sockaddr *)&cl_addr, &cl_len);
		if (connfd < 0) {
			perror(SELF ": accept()");
			return FAIL;
		}
		close(sockfd);
		sockfd = -1;
	}
	else /* if (con) */ { /* connect */
		if (connect(sockfd, (struct sockaddr *)&srv_addr,
					sizeof(srv_addr)) < 0) {
			perror(SELF ": connect()");
			return FAIL;
		}
		connfd = sockfd;
	}

	if (fd < 0) {
		/* redirect stdin/stdout */
		close(0);
		close(1);
		dup2(connfd, 0);
		dup2(connfd, 1);
		close(connfd);
	}
	else {
		if (fd != connfd) {
			close(fd); /* just in case */
			if (dup2(connfd, fd) != fd) {
				perror(SELF ": dup2()");
				return FAIL;
			}
			close(connfd);
		}
	}

	execvp(argv[optind], argv+optind);

	perror(SELF ": execvp()");
	return 127;
}
